src/coreclr/src/vm/peimagelayout.cpp

```diff
@@ -398,32 +398,89 @@ ConvertedImageLayout::ConvertedImageLayout(PEImageLayout* source)
     m_pOwner=source->m_pOwner;
     _ASSERTE(!source->IsMapped());
 
+    m_pExceptionDir = NULL;
+
     if (!source->HasNTHeaders())
         EEFileLoadException::Throw(GetPath(), COR_E_BADIMAGEFORMAT);
     LOG((LF_LOADER, LL_INFO100, "PEImage: Opening manually mapped stream\n"));
 
+#if !defined(CROSSGEN_COMPILE) && !defined(TARGET_UNIX)
+    // on Windows we may want to enable execution if the image contains R2R sections
+    // so must ensure the mapping is compatible with that
+    m_FileMap.Assign(WszCreateFileMapping(INVALID_HANDLE_VALUE, NULL,
+        PAGE_EXECUTE_READWRITE, 0,
+        source->GetVirtualSize(), NULL));
 
+    DWORD allAccess = FILE_MAP_EXECUTE | FILE_MAP_WRITE;
+#else
     m_FileMap.Assign(WszCreateFileMapping(INVALID_HANDLE_VALUE, NULL,
-                                          PAGE_READWRITE, 0,
-                                          source->GetVirtualSize(), NULL));
+        PAGE_READWRITE, 0,
+        source->GetVirtualSize(), NULL));
+
+    DWORD allAccess = FILE_MAP_ALL_ACCESS;
+#endif
+
     if (m_FileMap == NULL)
         ThrowLastError();
 
-
-    m_FileView.Assign(CLRMapViewOfFile(m_FileMap, FILE_MAP_ALL_ACCESS, 0, 0, 0,
+    m_FileView.Assign(CLRMapViewOfFile(m_FileMap, allAccess, 0, 0, 0,
                                 (void *) source->GetPreferredBase()));
     if (m_FileView == NULL)
-        m_FileView.Assign(CLRMapViewOfFile(m_FileMap, FILE_MAP_ALL_ACCESS, 0, 0, 0));
+        m_FileView.Assign(CLRMapViewOfFile(m_FileMap, allAccess, 0, 0, 0));
 
     if (m_FileView == NULL)
         ThrowLastError();
 
     source->LayoutILOnly(m_FileView, TRUE); //@TODO should be false for streams
     IfFailThrow(Init(m_FileView));
 
-#ifdef CROSSGEN_COMPILE
+#if defined(CROSSGEN_COMPILE)
     if (HasNativeHeader())
+    {
         ApplyBaseRelocations();
+    }
+#elif !defined(TARGET_UNIX)
+    if (HasCorHeader() && (HasNativeHeader() || HasReadyToRunHeader()) && g_fAllowNativeImages)
```

Comment:
Is this going to enable loading R2R images via e.g. `Load(byte[])` ? I am not sure whether it is something we want to enable.

Reply:
Is the concern that we would do redundant relocations or that the R2R code might be made to run via reflection?

Assuming that R2R code does the same thing as corresponding IL, I do not see problems with the latter. Perhaps even a good thing. 

That is - if that actually works. 
I did not think that `Load(byte[])` might hit this codepath. I will have to check what really happens.

Reply:
I checked `Load(byte[])` scenario and it appears to not using this codepath. `Load(byte[])` creates `RawImageLayout` and that does not come here.

Reply:
`ConvertedImageLayout` is used explicitly when loading from a Bundle creates that. 
There is also another codepath that could lead to creating `ConvertedImageLayout`. I am not sure if it is used and in what cases.

It should be possible to remember that the file was loaded form a bundle and perform the above just in those cases to be sure it does not enable more scenarios than needed.

Reply:
Added `m_isInBundle` to `ConvertedImageLayout` - to make sure this does not get enabled in other cases.

Reply:
In theory it is possible that `ConvertedImageLayout` could be used for other purposes, but even if it happens it seems to be rare. I have made a trial change where creating `ConvertedImageLayout` not for a purpose of a file in a bundle and the change passed all CI tests. (https://github.com/dotnet/runtime/pull/40403)

Anyways, we now should be relocating only when `ConvertedImageLayout` was created for a file in a bundle. If the other scenarios are reachable, they will still retain previous behavior.